tareas/Tarea3_Avd/optimizacion.jl

```diff
@@ -0,0 +1,1064 @@
+### A Pluto.jl notebook ###
+# v0.17.3
+
+using Markdown
+using InteractiveUtils
+
+# ╔═╡ a7cc31d3-e34a-425a-a698-e4e869c57a05
+using Plots
+
+# ╔═╡ 7295e189-d6f2-410b-bca0-597e1b4089ee
+include("intervalos.jl")
+
+# ╔═╡ 7dd4495e-73fc-11ec-37a7-9fbab661cf88
+#=function mini(f, I, tol=1e-3)   ### Non-Interval minimize method
+	y = f(I.infimo)
+	x = I.infimo
+	for i in (I.infimo+tol):tol:I.supremo
+		if f(i-tol) > f(i) < f(i+tol)
+			if f(i) < y
+				y = f(i)
+				x = i
+			end
+		end
+	end
+	return y, x
+end=#
+
+# ╔═╡ aab40414-6df7-4f1b-8a86-5d216ff3b61f
+#=function maxi(f, I, tol=1e-3)   ### Non-Interval maximize method
+	y = f(I.infimo)
+	x = I.infimo
+	for i in (I.infimo+tol):tol:I.supremo
+		if f(i-tol) < f(i) > f(i+tol)
+			if f(i) > y
+				y = f(i)
+				x = i
+			end
+		end
+	end
+	return y, x
+end=#
+
+# ╔═╡ 6811f544-94ba-4998-9ab7-0c156f042736
+#=function minimiza_no_peaks(f, I, tol)
+	(I.supremo - I.infimo) < tol && return I
+	I₀ = I
+	while true
+		mp = I₀.infimo + (I₀.supremo - I₀.infimo)/2
+		I₁ = Intervalo(I₀.infimo, mp);   I₂ = Intervalo(mp, I₀.supremo)
+		Ints = [I₁, I₂];   ys = []
+		for i in Ints
+			step = (i.supremo - i.infimo)/100
+			y_min = Inf
+			for p in i.infimo:step:i.supremo
+				if f(p-step) > f(p) < f(p+tol)
+					if f(p) < y_min
+						y_min = f(p)
+					end
+				end
+			end
+			push!(ys, y_min)
+		end
+		if ys[1] < ys[2]
+			I₀ = Ints[1]
+		elseif ys[2] < ys[1]
+			I₀ = Ints[2]
+		end
+		if (I₀.supremo - I₀.infimo) < tol
+			return I₀
+		end
+	end 
+end=#
+
+# ╔═╡ 5f4e4159-5517-4907-a583-3cdfbbcabbdb
+"""
+	minimiza(f, I, tol)
+Find the interval(s) (type=Intervalo) which contains the minimum of the function *f* in the initial given interval (type=Intervalo) *I*, until the diameter of the answer(s) meets the condition to have less length than *tol*.
+"""
+function minimiza(f, I, tol)
```

Comment:
```suggestion
function minimiza(f, I;tol=1e-2)
```

esto es opcional y solo un ejemplo de como puedes mantener tu función sin editarla, cuando pones el punto y coma le das a la función la opción de recibir o no la variable tol. Es decir, la implementación minimiza(f,I) funcionará sin necesidad de denifinir una tolerancia explicita pero bien se la puedes dar si tu lo deseas. Esta froma de definirla le da un valor fijo, así que sientete a gusto de cambiar ese 1e-2 que puse es solo de ejemplo y tu veras el más optimo y en caso de necesitar más tolerancia, entonces si puedes ingresar el valor explicitamente de la forma minimiza(f,I; I=1e-20)

Reply:
El comentario de @Julian-RC  es importante, ya que justo porque tu función define un método con tres parámetros, es que hay un error a la hora de correr los tests de la optimización. 

Puedes hacer dos cosas: (1) Como lo sugiere Julián, puedes definir la función con argumento que requiere un "keyword" (`tol` en este caso). (2) Puedes asignarle un valor default al tercer argumento (1e-2, o lo que quieras), que puedes cambiar, usando por ejemplo `minimiza(f, I, tol=1e-2)`. En ambos casos, obtendrás dos métodos para `minimiza`. Vale la pena hacer notar que el valor default debe ser suficientemente pequeño para que el test pase con el valor default que asignas; yo usé el valor `1/1024` que es un Float64 exactamente representable en base 2.